designs/package-spec.md

```diff
@@ -0,0 +1,78 @@
+# Package Controller Configuration
+
+## Introduction
+
+**Problem:** currently the ability to configure the package controller is very limited and requires code changes
+
+Code changes currently have to be made to configure the package controller during cluster creation or using the install command. Changes need to be coordinated between the CLI and the package controller helm chart to alter configuration. This makes it more difficult to support things like various private registry configurations.
+
+### Tenets
+
+* ***Flexible:*** reduce dependency of the CLI on helm chart.
+
+### Goals and Objectives
+
+As an EKS-A cluster administrator I want to configure the package controller for:
+
+* Harbor using proxy cache
+* Harbor using replication
+* Harbor as a private registry using the import images command
+* Configure ECR token refresher
+* Configure for kubelet private registry authentication
+* Support proxy configuration
+* Support package controller installation from alternate registries
+* Disable installation of package controller
+* Allow for future configuration options
+
+### Statement of Scope
+
+**In scope**
+* All of the above
+
+## Current state
+The package controller can be used currently with:
+1. Proxy configurations
+1. ECR token refresher enabled/disabled
+1. Installation of the package controller from alternate registries
+
+## Overview of Solution
+Add a section to the cluster specification to enable/disable the package controller installation and pass configuration values.
+
+## Solution Details
+
+Add a packageController section to the cluster specification for example to use development/staging builds:
+```yaml
+apiVersion: anywhere.eks.amazonaws.com/v1alpha1
+kind: Cluster
+metadata:
+  name: billy
+spec:
+  packageController:
```

Comment:
We probably need a validation to make sure this is only set for management clusters? Or would we allow to install it in workload clusters as well?

If it's allowed in workload clusters, we would need to document what happens when this is specified directly through the API, since we don't have a "packages controller operator" that would install/update the packages controller.

Reply:
These values potentially will be used for workload clusters as well. We currently just create some custom resources and secrets for workload clusters, but Vincent is working on a helm chart for them that would use this.

Reply:
Ah ok, that's where my packages knowledge falls short. Does that mean we are going to allow users to install the package controller in workload clusters as well?

I'm just trying to figure out what should be the behavior of this new field for workload clusters created through the controller and not through the CLI.

Reply:
The package controller will not be installed on workload clusters. A helm install will be done on the management cluster for each workload cluster so we can manage those configuration resources easier.

Reply:
Ah cool. So since this filed configures the package controller and the package controller is never installed in workload clusters, should we add a validation that checks that is always disabled and no extra configuration is provided for workload clusters?

Reply:
This configuration is used to configure the workload cluster *on* the management cluster for a workload cluster.

Reply:
Ok

Are all fields relevant for both management and workload cluster? When I look at the `controller` subfield, that seems like it only affects the controller deployment running in the management cluster?

What should the eks-a cluster controller do when a workload cluster is created through the API? does it need to install/update a helm chart?

Reply:
Yes, the controller section would only be relevant for a management cluster.

The cluster controller takes the same action for management and workload clusters, it runs the same helm install.